packages/lodestar/src/api/impl/beacon/state/utils.ts

```diff
@@ -37,8 +50,68 @@ export async function resolveStateId(
   //todo: resolve archive slot -> state
   const blockSummary = forkChoice.getCanonicalBlockSummaryAtSlot(slot);
   if (blockSummary) {
-    return (await db.stateCache.get(blockSummary.stateRoot))?.state ?? null;
+    return (await db.stateCache.get(blockSummary.stateRoot)) ?? null;
   }
 
   return null;
 }
+
+export function toValidatorResponse(index: ValidatorIndex, validator: Validator): ValidatorResponse {
+  return {
+    index,
+    status: ValidatorStatus.ACTIVE,
+    pubkey: validator.pubkey,
+    validator,
+  };
+}
+
+/**
+ * Returns committees mapped by index -> slot -> validator index
+ */
+export function getEpochBeaconCommittees(
+  config: IBeaconConfig,
+  chain: IBeaconChain,
+  stateContext: ApiStateContext,
+  epoch: Epoch
+): ValidatorIndex[][][] {
+  let committees: ValidatorIndex[][][] | null = null;
+  if (stateContext.epochCtx) {
+    switch (epoch) {
+      case chain.clock.currentEpoch: {
+        committees = stateContext.epochCtx.currentShuffling.committees;
+        break;
+      }
+      case chain.clock.currentEpoch - 1: {
+        committees = stateContext.epochCtx.previousShuffling.committees;
+        break;
+      }
+    }
+  }
+  if (!committees) {
+    const indicesBounded: [ValidatorIndex, Epoch, Epoch][] = readOnlyMap(stateContext.state.validators, (v, i) => [
+      i,
+      v.activationEpoch,
+      v.exitEpoch,
+    ]);
+
+    const shuffling = computeEpochShuffling(config, stateContext.state, indicesBounded, epoch);
+    committees = shuffling.committees;
+  }
+  return committees;
+}
+
+export function validatorPubkeyToIndex(
+  config: IBeaconConfig,
+  stateCtx: ApiStateContext,
+  pubkey: BLSPubkey
+): ValidatorIndex | null {
+  if (stateCtx.epochCtx) {
+    return stateCtx.epochCtx.pubkey2index.get(pubkey) ?? null;
+  }
+  for (const [index, validator] of Array.from(stateCtx.state.validators).entries()) {
```

Comment:
this place can potentially receive DOS attach. As head state/context is always available in cache, can we just use it?

Reply:
Oh true, validators are never deleted, so we could always use head epochCtx